### ISSUE TITLE
Monitoring outbound HTTPS external call

### ISSUE THREAD

**Issue by jmirc** [linkerd/linkerd2#2192]
## Feature Request
I would like Linkerd2 to monitor external service calls using HTTPS protocol. I would like to get at least the outbound info, the rps and the error rates.

### What problem are you trying to solve?
We are deploying our services under Kubernetes / EKS on AWS. We are using many AWS services and most of them are using HTTPS (SQS, SNS, etc...). A common pattern is the following:

                            
Service 1 ------> SQS -----> Service 2 -----> Service 3 (inside EKS)
                                                                   -----> Service 4 (outside EKS)

Only the communication between the Service 2 and the Service 3 will be monitored because they are running inside Kubernetes.

The communication between Service 2 and the Service 4 will not monitored because the communication uses HTTPS.

### How should the problem be solved?
N/A

What do you want to happen? Add any considered drawbacks.
N/A

### Any alternatives you've considered?
The Service1 could create a compatible metric to monitor outbound call
The Service 2 could create a compatible metric to monitor inbound call

The issue is related to grafana/prometheus because most of the graphics are bound to a deployment / pod and the external service doesn't have it. 

Is there another way to solve this problem that isn't as good a solution?

### How would users interact with this feature?

If you can, explain how users will be able to use this. Maybe some sample CLI
output?

**Comment by jmirc** [comment]
Another solution, could be to offer a rest api to interact with the sidecar. The api could offer the capability to 
  - increment outbound counter (request_total, response_total)
  - send the response_latency timer

**Comment by jmirc** [comment]
@grampelberg I will prefer using the first solution even if it is more work for the first integration, it will be transparent for the client if we are using or not the proxy.

**Comment by bdimcheff** [comment]
We're also looking to monitor a ~dozen external service calls, all over https. I had tried to get this working with linkerd for a couple hours back in December, not realizing that TLS was likely my problem. 

Is it possible for linkerd to mitm outgoing https connections in such a way that per-endpoint config (besides trusting linkerd's cert) isn't necessary? I think trusting a cert/CA is probably an ok compromise for our purposes. If linkerd's control plane generates its cert as a secret, it can probably be mounted into the app's container in the right place for validation to succeed. Or we can statically generate a cert and stick it in our CI build somewhere.

In our case, it'd also be possible for us to specify a different url for our integrations if necessary. Every one of the libraries we use (so far) allow for a custom urls to be provided. It'd just be less friendly that way, since we'd have to maintain a fair amount of config. If that's what we do, I don't know that I particularly care about TLS between the app container and the sidecar linkerd.

Regardless of what happens, it'd be extremely important to me that linkerd would verify https certs for the connections it makes. I'd like failures to be fatal (like send a 500 back to my app) since I really don't want to be accidentally making insecure connections across the public internet.

**Comment by grampelberg** [comment]
@bdimcheff that's really useful. Sounds like it is worth looking into the MITM config a little bit more.

**Comment by bdimcheff** [comment]
Anyone know what [istio is doing](https://istio.io/docs/tasks/traffic-management/egress/#configuring-istio-external-services) to deal with tls?

**Comment by grampelberg** [comment]
It appears that they require [configuration](https://istio.io/docs/examples/advanced-egress/egress-tls-origination/#tls-origination-for-egress-traffic). My understanding is that istio's egress is opt-in (eg all traffic is blocked by default), so that makes some sense.

**Comment by StevenACoffman** [comment]
It looks like they are moving to using their own CNI: https://github.com/istio/cni

> For application pods in the Istio service mesh, all traffic to/from the pods needs to go through the sidecar proxies (istio-proxy containers). This istio-cni CNI plugin will set up the pods' networking to fulfill this requirement in place of the current Istio injected pod initContainers istio-init approach.
> 
> This is currently accomplished (for IPv4) via configuring the iptables rules in the netns for the pods.
> 
> The CNI handling the netns setup replaces the current Istio approach using a NET_ADMIN privileged initContainers container, istio-init, injected in the pods along with istio-proxy sidecars. This removes the need for a privileged, NET_ADMIN container in the Istio users' application pods.

**Comment by bdimcheff** [comment]
yeah i think what's interesting about istio's docs above is that (in their example) it allows you to exec into the pod and then `curl https://www.google.com`, presumably without getting a cert error. That config must somehow configure a mitm, no?

fwiw it'd definitely meet our needs if we needed to define service profiles (or some other k8s crd config) for each external TLS service that we wanted linkerd to intercept. I'd like to avoid having to manually alter a bunch of config on our service itself, but linkerd config is totally reasonable/expected.

**Comment by grampelberg** [comment]
@bdimcheff as I understand the docs a `curl https://www.google.com` without a VirtualService will not  be mitm'd or have any stats. I *tried* to figure out the egress prometheus metrics and gave up though, so I might be entirely wrong.

I still really like the idea of mitm, especially if the injector can put trust bundles in the correct place automatically. We'll just need to figure out how feasible it is. This crazy amount of config frightens me.

**Comment by bdimcheff** [comment]
> as I understand the docs a `curl https://www.google.com` without a VirtualService will not be mitm'd or have any stats.

yeah that's my understanding as well, which I think is fine/expected. I appreciate the ways in which linkerd just magically works, but in this case I'm definitely willing to write some config. There may be some outgoing TLS connections that I legitimately _don't_ want linkerd to intercept, and it seems reasonable to me for TLS MITM to be opt-in.

**Comment by darthmolen** [comment]
We have a use case where there are a bunch of 3rd party integrations we connect to and would love to tap, monitor, and metric this traffic using linkerd.

**Comment by eloo-abi** [comment]
Second that, 

is this feature currently planned?

**Comment by channyein87** [comment]
Really need this feature to monitor the external traffic from the pod for observibility.

**Comment by ryanisnan** [comment]
This is probably fixed in Linkerd 2.17

**Comment by alpeb** [comment]
Thanks for the flagging this one out. Yes, this is largely covered by the EgressNetwork functionality introduced in 2.17, paired with TLSRoute resources. There's other related functionality that will get covered via TLS origination that is currently being worked on, but for now let's close this one out, and please open separate ones if there are more specific issues we should take into consideration.
